File: LoadLink-BE/app/routes/review.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from datetime import date

from app import models
from app.schemas.review import ReviewCreate, ReviewUpdate, ReviewResponse
from app.dependencies import get_db, get_current_user

router = APIRouter(prefix="/reviews", tags=["Reviews"])

# ------------------
# Create Review
# ------------------
@router.post("/", response_model=ReviewResponse)
def create_review(
    review_in: ReviewCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    # Check if booking exists
    booking = db.query(models.Booking).filter_by(id=review_in.booking_id).first()
    if not booking:
        raise HTTPException(status_code=404, detail="Booking not found")

    # Prevent user from reviewing themselves
    if review_in.to_user_id == current_user.id:
        raise HTTPException(status_code=400, detail="Cannot review yourself")

    review = models.Review(
        booking_id=review_in.booking_id,
        from_user_id=current_user.id,
        to_user_id=review_in.to_user_id,
        rating=review_in.rating,
        comment=review_in.comment,
        created_date=date.today()
    )

    db.add(review)

    # ------------------
    # Update booking review flags
    # ------------------
    if current_user.id == booking.shipper_id:
        booking.shipper_reviewed = True
    elif current_user.id == booking.trip.carrier_id:  # Assuming trip has carrier_id
        booking.carrier_reviewed = True

    db.commit()
    db.refresh(review)
    return review


# ------------------
# Get All Reviews
# ------------------
@router.get("/", response_model=list[ReviewResponse])
def get_reviews(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    return db.query(models.Review).all()

# ------------------
# Get Review by ID
# ------------------
@router.get("/{review_id}", response_model=ReviewResponse)
def get_review(
    review_id: str,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    review = db.query(models.Review).filter_by(id=review_id).first()
    if not review:
        raise HTTPException(status_code=404, detail="Review not found")
    return review

# ------------------
# Update Review
# ------------------
@router.put("/{review_id}", response_model=ReviewResponse)
def update_review(
    review_id: str,
    review_in: ReviewUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    review = db.query(models.Review).filter_by(id=review_id).first()
    if not review:
        raise HTTPException(status_code=404, detail="Review not found")

    # Only author can update
    if review.from_user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    for key, value in review_in.dict(exclude_unset=True).items():
        setattr(review, key, value)

    db.commit()
    db.refresh(review)
    return review

# ------------------
# Delete Review
# ------------------
@router.delete("/{review_id}", status_code=204)
def delete_review(
    review_id: str,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    review = db.query(models.Review).filter_by(id=review_id).first()
    if not review:
        raise HTTPException(status_code=404, detail="Review not found")

    # Only author can delete
    if review.from_user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    db.delete(review)
    db.commit()
    return

File: LoadLink-FE/services/review.ts
"use client";

import api from "../lib/api";

// ------------------
// Types
// ------------------
export interface ReviewBase {
  booking_id: string;
  to_user_id: string;
  rating: number; // 1-5
  comment?: string;
}

export interface ReviewCreate extends ReviewBase {}

export interface ReviewUpdate {
  rating?: number;
  comment?: string;
}

export interface ReviewOut extends ReviewBase {
  id: string;
  from_user_id: string;
  created_date: string; // ISO string
}

// ------------------
// API calls
// ------------------

// Create a review
export const createReviewApi = async (
  reviewData: ReviewCreate
): Promise<ReviewOut> => {
  const res = await api.post("/reviews/", reviewData);
  return res.data;
};

// Get all reviews
export const getReviewsApi = async (): Promise<ReviewOut[]> => {
  const res = await api.get("/reviews/");
  return res.data;
};

// Get review by ID
export const getReviewByIdApi = async (
  reviewId: string
): Promise<ReviewOut> => {
  const res = await api.get(`/reviews/${reviewId}`);
  return res.data;
};

// Update review
export const updateReviewApi = async (
  reviewId: string,
  reviewData: ReviewUpdate
): Promise<ReviewOut> => {
  const res = await api.put(`/reviews/${reviewId}`, reviewData);
  return res.data;
};

// Delete review
export const deleteReviewApi = async (reviewId: string): Promise<void> => {
  await api.delete(`/reviews/${reviewId}`);
};
